### PR TITLE
Fixes Issue 186: specify custom directories for rhel secrets

### LINKED ISSUE #186: Need a way to specify custom directories for rhsm configuration
Patch https://github.com/docker/docker/pull/6075 add support to docker to specify "secret" files to be injected in a container. As of today's Atomic Hosts when one does ```docker run -it registry.access.redhat.com/rhel7``` the secrets files are copied over and injected into the container from ```/usr/share/rhel/secrets```. This enables subscription manager to run inside containers as desired.

On the atomic host we see the forllowing 
```
$ ls -al /usr/share/rhel/secrets/
lrwxrwxrwx. 1 root root 20 Jul 28 01:00 etc-pki-entitlement -> /etc/pki/entitlement          
lrwxrwxrwx. 1 root root 28 Jul 28 01:00 rhel7.repo -> /etc/yum.repos.d/redhat.repo           
lrwxrwxrwx. 1 root root  9 Jul 28 01:00 rhsm -> /etc/rhsm                                    
```

So consider the case where you want the /etc/rhsm on the host to be different from the /etc/rhsm inside the container. It could be something basic like "the rhsm base url for yum repos is different from base url for atomic ostree repos" OR "we want to use different entitlement models and hide access to certain repos".  The above configuration will make it impossible to change that because ```/usr/share/rhel/secrets/rhsm``` is a read-only link . The symlink there cannot be redirected.
The ```/etc/rhsm``` conf symlinked above specifically points hosts rhsm conf. We cannot have separate configurations for hosts and containers.

In short we need something like this 

```
$ ls -al /usr/share/rhel/secrets/
lrwxrwxrwx. 1 root root 20 Jul 28 01:00 etc-pki-entitlement -> /etc/pki/entitlement-latest          
lrwxrwxrwx. 1 root root 28 Jul 28 01:00 rhel7.repo -> /etc/yum.repos.d/redhat-latest.repo           
lrwxrwxrwx. 1 root root  9 Jul 28 01:00 rhsm -> /etc/rhsm-latest

$ ls -al /etc/pki/entitlement-latest
entitlement-latest -> entitlements

$ ls -al /etc/rhsm-latest
rhsm-latest -> rhsm

$ ls -al /etc/yum.repos.d/redhat-latest.repo
redhat-latest.repo -> redhat.repo                                   
```
What this would enable is for one to change what /etc/rhsm-latest points to and specify an alternate configuration for the container while preserving the default behavior when desired.

### PR DESCRIPTION
**Pull request by shishir-a412ed**
Fixes #186 
Signed-off-by: Shishir Mahajan <shishir.mahajan@redhat.com>

### CODE REVIEW

**Comment by shishir-a412ed** [comment]
Carryover of #187 
ping @rhatdan @runcom

**Comment by runcom** [comment]
Mmm I think this isnt just a carry, instead, this PR should take care of continuing from https://github.com/projectatomic/docker/pull/187#discussion_r76056631and implement what Dan and Colin was saying..

File: container/secrets.go [hunk 1]
```diff
@@ -81,5 +81,13 @@ func readFile(root, name string) ([]SecretData, error) {
 }
 
 func getHostSecretData() ([]SecretData, error) {
-	return readAll("/usr/share/rhel/secrets", "")
+	baseSecrets, err := readAll("/usr/share/rhel/secrets", "")
+	if err != nil {
+		return nil, err
+	}
+	overrideSecrets, err := readAll("/etc/container/rhel/secrets", "")
```
Comment (rhatdan) [inline container/secrets.go]: If the directory does not exist, we should not fail.

**Comment by rhatdan** [comment]
We probably want to merge this patch into the secrets patch once it is approved.

File: container/secrets.go [hunk 2]
```diff
@@ -81,5 +81,13 @@ func readFile(root, name string) ([]SecretData, error) {
...
+	if err != nil {
+		return nil, err
+	}
+	return append(baseSecrets, overrideSecrets...), nil
```
Comment (runcom) [inline container/secrets.go]: Does this take care of merging the two directories?

Comment (runcom) [inline container/secrets.go]: I mean, when integrating on the slice later in the code...

File: container/secrets.go (hunk 1 above)
Comment (rhatdan) [inline container/secrets.go]: I think we wanted just /etc/containers/secrets also, no need to hard code rhel.
/etc/containers/secrets
Note plural, which I think we are using for /var/lib/containers

@cgwalters WDYT

Comment (runcom) [inline container/secrets.go]: Correct, /etc/containers/secrets

Comment (shishir-a412ed) [inline container/secrets.go]: This does not fail when the directory does not exist.
e.g if `/etc/container/rhel/secrets` does not exist, then it is just gonna return a slice with `baseSecrets` and vice versa.

File: container/secrets.go (hunk 2 above)
Comment (shishir-a412ed) [inline container/secrets.go]: yes it works. If both of the directories exists, then in the container rootfs, you will eventually have the union of files in PATH1 and PATH2. With `/etc/container/rhel/secrets` overriding `/usr/share/rhel/secrets`.

**Comment by shishir-a412ed** [comment]
@rhatdan Updated the override path 
FROM: `/etc/container/rhel/secrets`
TO: `/etc/containers/secrets`

File: container/secrets.go [hunk 3]
```diff
@@ -81,5 +81,13 @@ func readFile(root, name string) ([]SecretData, error) {
 }
 
 func getHostSecretData() ([]SecretData, error) {
-	return readAll("/usr/share/rhel/secrets", "")
+	baseSecrets, err := readAll("/usr/share/rhel/secrets", "")
+	if err != nil {
+		return nil, err
```
Comment (rhatdan) [inline container/secrets.go]: Could we add a little documentation to this error.

return nil, fmt.Errorf("Failed to read /usr/share/rhel/secrets: %s", err.Error())

**Comment by shishir-a412ed** [comment]
@rhatdan PTAL

**Comment by runcom** [comment]
LGTM

ping @rhatdan @cgwalters

**Comment by runcom** [comment]
we need this patch fixed in fedora-1.10.3, rhel7-1.10.3, and docker-1.13 also

**Comment by rhatdan** [comment]
LGTM

**Comment by rhatdan** [comment]
Lokesh can you merge this with the secrets patch so we have one less patch.

**Comment by runcom** [comment]
@rhatdan, @lsm5 please do not. The branch docker-1.12 is a freezed branch since upstream 1.12 is GA, we will lose commit history if we squash these changes with the secret patch. The only branch this is going to be squashed is docker-1.13 because upstream hasn't related it yet. Older branch are in a maintaining mode where patches are only added and we never lose commit history by squashing. 

I'll take care of squashing these changes in docker-1.13, fedora-1.10.3 and rhel-1.10.3 need just a PR to have this changes applied as well.

**Comment by rhatdan** [comment]
SGTM.  As long as the list of patches we are carrying at projectatomic.io does not grow for the master branch, I am happy

**Comment by runcom** [comment]
Exactly, the patches in the development branch (in this case docker-1.13) never grow in numbers, that's where I squash commits because we can afford to lose git history from commit to commit.

File: container/secrets.go [hunk 4]
```diff
@@ -81,5 +82,16 @@ func readFile(root, name string) ([]SecretData, error) {
 }
 
 func getHostSecretData() ([]SecretData, error) {
-	return readAll("/usr/share/rhel/secrets", "")
+	baseDir := "/usr/share/rhel/secrets"
+	overrideDir := "/etc/containers/secrets"
```
Comment (runcom) [inline container/secrets.go]: `baseDir` was already defined at the top of this file and `overrideDir` could have been defined there as well saving those local vars

**Pull request merged**